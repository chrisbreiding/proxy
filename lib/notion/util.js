const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
