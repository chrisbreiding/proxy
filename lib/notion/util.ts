import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
