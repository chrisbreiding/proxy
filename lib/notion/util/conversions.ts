import type { BlockObjectResponse, RichTextItemResponse } from '@notionhq/client/build/src/api-endpoints'
import type { Block, NotionBlock, OutgoingBlock } from '../types'
import { clone } from '../../util/collections'

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type
  // @ts-ignore
  const content = block[type] as Content

  return {
    content: clone(content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}
