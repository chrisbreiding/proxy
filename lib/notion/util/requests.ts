import type { Method } from 'axios'
import { request } from '../../util/network'

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

export function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}
