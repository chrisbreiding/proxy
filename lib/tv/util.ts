import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
