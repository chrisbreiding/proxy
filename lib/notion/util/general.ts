import type express from 'express'
import type {
  ApiColor,
  BlockObjectResponse,
  RichTextItemResponse,
} from '@notionhq/client/build/src/api-endpoints'

import type { Block, Content, NotionBlock, OwnBlock } from '../types'
import { richTextToPlainText } from './conversions'

export function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  content?: Content
  text?: string
  type: BlockObjectResponse['type']
}

export function makeBlock (options: MakeBlockOptions) {
  const { annotations, children, content, text, type = 'paragraph' } = options

  return {
    type,
    children,
    content: content || {
      rich_text: typeof text === 'string' ? [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ] : [],
    },
  } as OwnBlock
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

// compare guids without dashes in case one does not include dashes
export function areIdsEqual (id1: string, id2: string) {
  return id1.replaceAll('-', '') === id2.replaceAll('-', '')
}

function getBlockChildrenDepth (block: Block, depth: number): number {
  if (!block.children) {
    return depth
  }

  return getBlocksChildrenDepth(block.children, depth + 1)
}

export function getBlocksChildrenDepth (blocks: Block[], depth = 0) {
  if (!blocks.length) return depth

  return Math.max(...blocks.map((block) => getBlockChildrenDepth(block, depth)))
}

export function isChildPageWithTitle (block: Block, titleOrComparison: string | ((title: string) => boolean)) {
  return (
    block.type === 'child_page'
    && 'title' in block.content
    && (
      typeof titleOrComparison === 'string'
        ? block.content.title === titleOrComparison
        : titleOrComparison(block.content.title)
    )
  )
}

export function sendHtml (res: express.Response, statusCode: number, message: string) {
  res
  .status(statusCode)
  .set('Content-Type', 'text/html')
  .send(message)
}

function htmlErrorPart (label: string, value: string) {
  if (!value) return ''

  return `
    <p>${label}</p>
    <pre>${value}</pre>
  `
}

interface HtmlErrorOptions {
  error: any
  message: string
  res: express.Response
  statusCode: number
}

export function sendHtmlError ({ error, message, res, statusCode }: HtmlErrorOptions) {
  sendHtml(res, statusCode, `
    <!DOCTYPE html>
    <html>
      <head>
        <title>⚠️ Error!</title>
        <style>
          body {
            padding: 20px;
          }
          pre {
            background: #f8f8f8;
            border: solid 1px #d7d7d7;
            overflow: auto;
            padding: 10px;
          }
        </style>
      </head>
      <body>
        <h3>${message}</h3>
        ${htmlErrorPart('Code', error?.code)}
        ${htmlErrorPart('Message', error?.message)}
        ${htmlErrorPart('Stack', error?.stack)}
        ${htmlErrorPart('Data code', error?.response?.data?.code)}
        ${htmlErrorPart('Data status', error?.response?.data?.status)}
        ${htmlErrorPart('Data message', error?.response?.data?.message)}
      </body>
    </html>
  `)
}
