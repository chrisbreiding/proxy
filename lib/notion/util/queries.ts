import type {
  BlockObjectResponse,
  DatabaseObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  QueryDatabaseBodyParameters,
} from '@notionhq/client/build/src/api-endpoints'

import { convertNotionBlockToOwnBlock } from './conversions'
import { makeRequest } from './requests'
import type { NotionBlock } from '../types'
import { compact } from '../../util/collections'

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({
    notionToken,
    path: `blocks/${pageId}/children`,
  }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

interface GetColumnBlocksOptions {
  columnListId: string
  notionToken: string
}

export async function getColumnBlocks ({ columnListId, notionToken }: GetColumnBlocksOptions) {
  const columns = await getBlockChildren({ notionToken, pageId: columnListId })
  const columnsBlocks = await Promise.all(columns.map(async (column) => {
    return getBlockChildrenDeep({ notionToken, pageId: column.id })
  }))

  return columnsBlocks.flat()
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
  filter?: QueryDatabaseBodyParameters['filter']
  startCursor?: QueryDatabaseBodyParameters['start_cursor']
}

export function queryDatabases ({ notionToken, databaseId, filter, startCursor }: QueryDatabasesOptions) {
  const body: QueryDatabaseBodyParameters = {}

  if (filter) {
    body.filter = filter
  }

  if (startCursor) {
    body.start_cursor = startCursor
  }

  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
    body: (filter || startCursor) ? body : undefined,
  })
}

interface GetDatabasePagesOptions {
  databaseId: string
  notionToken: string
  results?: DatabaseObjectResponse[]
  startCursor?: string
}

export async function getDatabasePages (options: GetDatabasePagesOptions): Promise<DatabaseObjectResponse[]> {
  const { databaseId, notionToken, startCursor } = options

  const { next_cursor, has_more, results } = await queryDatabases({
    databaseId,
    notionToken,
    startCursor,
  })

  if (!has_more) {
    return (options.results || []).concat(results as DatabaseObjectResponse[])
  }

  return getDatabasePages({
    databaseId,
    notionToken,
    results: results as DatabaseObjectResponse[],
    startCursor: next_cursor!,
  })
}
