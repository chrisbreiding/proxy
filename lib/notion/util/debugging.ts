import { getBlockPlainText, getRichText } from './general'
import { debug } from '../../util/debug'
import type { NotionBlock } from '../types'

const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
