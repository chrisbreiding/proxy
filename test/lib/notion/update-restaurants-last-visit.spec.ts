import { describe, it } from 'vitest'

import { block, nockGetBlockChildren, nockNotion } from '../../support/util'
import restaurantsLastVisit from '../../../lib/notion/update-restaurants-last-visit'

describe('lib/notion/update-restaurants-last-visit', () => {
  it('updates changed restaurant last visit dates', async () => {
    function blocksWithDate (date) {
      return {
        results: [
          block({ text: 'Note: something to note' }),
          block({ text: '' }),
          block({ text: date }),
          block({ text: 'Things we ate', type: 'bulleted_list_item' }),
          block({ text: 'And what we thought about it', type: 'bulleted_list_item' }),
        ],
      }
    }

    function bodyWithDate (date) {
      return {
        properties: {
          'Last Visit': {
            date: {
              start: date,
            },
          },
        },
      }
    }

    nockNotion({
      method: 'post',
      path: '/v1/databases/restaurants-id/query',
      fixture: 'restaurants/restaurant-pages',
    })

    nockGetBlockChildren('restaurant-1', { reply: blocksWithDate('9/7/22') })
    nockGetBlockChildren('restaurant-2', { reply: blocksWithDate('11/2/22') })
    nockGetBlockChildren('restaurant-3', { reply: blocksWithDate('7/29/22') })
    nockGetBlockChildren('restaurant-4', { reply: blocksWithDate('12/20/22') })
    nockGetBlockChildren('restaurant-5', { reply: blocksWithDate('6/10/22') })

    nockNotion({
      body: bodyWithDate('2022-11-02'),
      method: 'patch',
      path: '/v1/pages/restaurant-2',
    })

    nockNotion({
      body: bodyWithDate('2022-12-20'),
      method: 'patch',
      path: '/v1/pages/restaurant-4',
    })

    await restaurantsLastVisit.updateRestaurantsLastVisit({
      notionToken: 'notion-token',
      restaurantsDatabaseId: 'restaurants-id',
    })
  })
})
