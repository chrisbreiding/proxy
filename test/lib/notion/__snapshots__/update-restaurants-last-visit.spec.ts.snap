// Vitest Snapshot v1

exports[`lib/notion/update-restaurants-last-visit > updates changed restaurant last visit dates > restaurant-2 1`] = `
{
  "properties": {
    "Last Visit": {
      "date": {
        "start": "2022-11-02",
      },
    },
  },
}
`;

exports[`lib/notion/update-restaurants-last-visit > updates changed restaurant last visit dates > restaurant-4 1`] = `
{
  "properties": {
    "Last Visit": {
      "date": {
        "start": "2022-12-20",
      },
    },
  },
}
`;
